subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingTaskOperationListener.java

```diff
@@ -65,21 +62,21 @@
 class ClientForwardingTaskOperationListener extends SubtreeFilteringBuildOperationListener<ExecuteTaskBuildOperationDetails> {
 
     private final Map<TaskIdentity<?>, DefaultTaskDescriptor> descriptors = new ConcurrentHashMap<>();
-    private final Map<Path, TaskExecutionGraph> taskExecutionGraphs = new HashMap<>();
     private final OperationResultPostProcessor operationResultPostProcessor;
+    private final TaskExecutionGraphTracker taskExecutionGraphTracker;
     private final TaskOriginTracker taskOriginTracker;
 
+    @Nullable
+    private final TransformOperationTracker transformOperationTracker;
```

Comment:
Why is this nullable?

Reply:
Because we pass `null` to the constructor if transform operation events have not been requested.

Reply:
My preference would be the null object pattern in that case.

Reply:
Changed in e85fb9a33b2.